### PR TITLE
Microverse Module 1 Portfolio pair programming project: Data store in browser

### PR DESCRIPTION
**Pull request by binyamolango**
# Hello Reviewer 👋

> Here are the changes that we implemented

- Created a single JavaScript object with all the data from the entire form and saved it in local storage.
- When the user changes the content of any input field, the data is saved to the local storage.
- When the user loads the page, if there is any data in the local storage the input fields are pre-filled with this data.

### CODE REVIEW

File: script-preserveData.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  localStorage.setItem('form', JSON.stringify(data));
+}
+
+form.addEventListener('input', storeData);
+
+const formObj = JSON.parse(localStorage.getItem('form'));
+
+document.getElementById('name').value = formObj.name;
+document.getElementById('email').value = formObj.email;
+document.getElementById('msg').value = formObj.text;
```
Comment (Alusp) [inline script-preserveData.js]: - [ ] Good job far, :clap: However, Your form data send data using `formspree` but does not save data in the broswer's `Localstorage`, and also, your page does not persist data on the form when `refreshed or reloaded`

Please check your `script-preservedData.js` file and ensure that it is imported correctly in your `index.html` file :slightly_smiling_face: 

Below is a screenshot of your site: :arrow_down: 

Student Data

**Review by Alusp — CHANGES_REQUESTED** [review]
Hi @binyamolango, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
 
close

## Highlights
- [ ] Good Commit History :green_circle: 
- [ ] Good Github Flow :green_circle: 
- [ ] Good PR Summary :green_circle: 
- [ ] No Linter Errors :green_circle: 


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
 
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: script-validation.js [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
     form.submit();
   } else {
     e.preventDefault();
     emailErr.textContent = 'Please write your email in lower case!';
   }
-});
\ No newline at end of file
+}
+
+form.addEventListener('submit', emailValidation);
```
Comment (Cabraham1) [inline script-validation.js]: - [ ] Good job so far! However, your form data is sent using Formspree but does not save data in the browser's LocalStorage. Additionally, your page does not persist data on the form when refreshed or reloaded.

To address these issues, please check your script-preservedData.js file and ensure that it is imported correctly in your index.html file.

I would also like to suggest using LocalStorage to save the form data in the browser. By utilizing LocalStorage, you can store the data locally and retrieve it even after the page is refreshed or reloaded.

To implement this, you can modify your script-preservedData.js file to include the necessary code for interacting with LocalStorage. You can use the localStorage.setItem() method to store the form data and localStorage.getItem() method to retrieve it.

Here's an example of how you can save the form data to LocalStorage:

Retrieve form data
const formData = {
  // Your form data here
};

Save form data to LocalStorage
localStorage.setItem('formData', JSON.stringify(formData));

And here's how you can retrieve the saved data from LocalStorage:

Retrieve form data from LocalStorage
const savedData = localStorage.getItem('formData');

if (savedData) {
  const formData = JSON.parse(savedData);
  // Populate the form fields with formData
}

By implementing LocalStorage in your script, you can ensure that the form data persists across page refreshes or reloads.

**Review by Cabraham1 — CHANGES_REQUESTED** [review]
# Required Changes :recycle:
almost-there

## Hey @binyamolango   :wave:

## Highlights :1st_place_medal:


✔️ Following Gitflow
✔️ Following some requirements

Good job on implementing some of the requirements, there are some issues that you still need to work on to go to the next project but you are almost there! :blush:

# Optional

_All the changes with optional tags are not crucial enough to prevent you from getting the approval but we highly recommend to students that they should take these changes into consideration._

Cheers, and Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

---------------------------------------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by brainconnect93 — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi @binyamolango, 👋 👋  

Your project is complete! Great job for following the project requirement🥇and implementing the changes requested by the previous reviewer. There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 




## To Highlight 👍

- Descriptive Pull Request title and summary. ✔️
- Preserve data in the LocalStorage ✔️ 
- Good use of Github flow ✔️

## Optional changes

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @brainconnect93 in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 2 more limited reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**